io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs

```diff
@@ -0,0 +1,176 @@
+              g w' <$> runWriterT (release resource (ExitCaseSuccess b)) 
+            ExitCaseException err ->
+              g w  <$> runWriterT (release resource (ExitCaseException err))
+            ExitCaseAbort ->
+              g w  <$> runWriterT (release resource ExitCaseAbort))
+        (\(resource, w)   -> g w <$> runWriterT (use resource))
+    where f ((x,_),(y,w)) = ((x,y),w)
+          g w (a,w') = (a,w<>w')
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadMask m) => MonadMask (WriterT w m) where
+  mask f = WriterT $ mask $ \u -> runWriterT $ f (q u)
+    where
+      q :: (forall x. m x -> m x)
+        -> WriterT w m a -> WriterT w m a
+      q u (WriterT b) = WriterT (u b)
+  uninterruptibleMask f = WriterT $ uninterruptibleMask $ \u -> runWriterT $ f (q u)
+    where
+      q :: (forall x. m x -> m x)
+        -> WriterT w m a -> WriterT w m a
+      q u (WriterT b) = WriterT (u b)
+
+
+--
+-- RWST Instances
+--
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadThrow (RWST r w s m) where
+  throwIO = lift . throwIO
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadCatch (RWST r w s m) where
+  catch (RWST m) f = RWST $ \r s -> catch (m r s) (\e -> runRWST (f e) r s)
+
+  -- | general bracket ignores the state produced by the release callback
+  generalBracket acquire release use = RWST $ \r s ->
+      f <$> generalBracket
+              (runRWST acquire r s)
+              (\(resource, s', w') e ->
+                case e of
+                  ExitCaseSuccess (b, s'', w'') ->
+                    g w'' <$> runRWST (release resource (ExitCaseSuccess b)) r s''
+                  ExitCaseException err ->
+                    g w'  <$> runRWST (release resource (ExitCaseException err)) r s'
+                  ExitCaseAbort ->
+                    g w'  <$> runRWST (release resource  ExitCaseAbort) r s')
+              (\(a, s', w')   -> g w' <$> runRWST (use a) r s')
+    where
+      f ((x,_,_),(y,s,w)) = ((x,y),s,w)
+      g w (x,s,w') = (x,s,w<>w')
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadMask m) => MonadMask (RWST r w s m) where
+  mask f = RWST $ \r s -> mask $ \u -> runRWST (f (q u)) r s
+    where
+      q :: (forall x. m x -> m x)
+        -> RWST r w s m a -> RWST r w s m a
+      q u (RWST b) = RWST $ \r s -> u (b r s)
+  uninterruptibleMask f = RWST $ \r s -> uninterruptibleMask $ \u -> runRWST (f (q u)) r s
+    where
+      q :: (forall x. m x -> m x)
+        -> RWST r w s m a -> RWST r w s m a
+      q u (RWST b) = RWST $ \r s -> u (b r s)
+
+--
+-- StateT instances
+--
+
+-- | @since 1.0.0.0
+instance MonadCatch m => MonadThrow (StateT s m) where
+  throwIO = lift . throwIO
+
+-- | @since 1.0.0.0
+instance MonadCatch m => MonadCatch (StateT s m) where
+  catch (StateT m) f = StateT $ \s -> catch (m s) (\e -> runStateT (f e) s)
+
+  -- | general bracket ignores the state produced by the release callback
+  generalBracket acquire release use = StateT $ \s -> fmap f $
```

Comment:
The same as https://hackage.haskell.org/package/exceptions-0.10.7/docs/src/Control.Monad.Catch.html#local-6989586621679055622